### PR TITLE
Introduce public storage class for extending

### PR DESCRIPTION
**Pull request by PVince81**
## Description
When writing external storage it is complicated to reimplement IStorage
and many others. This PR provides public classes to be extended to
provider storage specific implementations.

## Related Issue
Because apps need to be able to extend the common storage class or use the flysystem adapter without accessing the private namespaces.

Also see https://github.com/owncloud/documentation/pull/2721

## How Has This Been Tested?
- [x] TODO: use it in files_external_ftp

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

@DeepDiver1975 @jvillafanez @labkode FYI

### CODE REVIEW

**Comment by PVince81** [comment]
Tested with this PR https://github.com/owncloud/files_external_ftp/pull/11 from the FTP app.

To make it a bit better I'll also make the storages from files_external extend that one.

**Comment by PVince81** [comment]
Ok, done. The storages from files_external now extend `OCP\Files\Storage\StorageAdapter`.
I've tested with Dropbox and it still works.

**Comment by jvillafanez** [comment]
We should define the abstract methods required to be implemented, first in OC\Files\Storage\Common and then in the public adapter. That way is way easier for anyone to know what he should implement.

It's **REALLY** annoying creating a new subclass, implementing a couple of methods and when you consider it's finally ready then you notice that an additional method should be implemented because the class explodes.

**Comment by PVince81** [comment]
@jvillafanez do you mean checking all abstract methods from the interface and redefine these as abstract within the StorageAdapter class ?

Indeed, the Common class doesn't implement ALL the methods, there are still a few remaining... (me really longing for a better simplified storage API in the future)

**Comment by jvillafanez** [comment]
Yes. In case of the `stat` method, you have to check the new public class (not defined there), check the Common class (again not defined there) and finally the interface. The same happens with the `getId` method. Those are a lot of "jumps" to know what are the methods that should be implemented.

There is also a weird inheritance chain: if you want to implement the new public storage, you have to check the Common class, which is private, and then the interface, which is public again. Maybe the Common class should be moved to the public library....

I think it would be better if the abstract classes define the list of abstract methods and they also document how they should be implemented. That way, just looking at the class, you have all the information you need to implement the class.

**Comment by PVince81** [comment]
> I think it would be better if the abstract classes define the list of abstract methods and they also document how they should be implemented. That way, just looking at the class, you have all the information you need to implement the class.

Ok, I'll do that short term.

Middle to long term we should probably move all methods from `Common` into `StorageAdapter` and make every other storage, even the internal ones, extend from it.

**Comment by PVince81** [comment]
@jvillafanez I've added the abstract methods that a storage must implement in `StorageAdapter` and `FlysystemStorageAdapter`.

**Comment by PVince81** [comment]
Okay, the code checker wants me to add `@since` tags for this "new" public APIs, will do...

**Comment by jvillafanez** [comment]
The only thing I miss is the expected exceptions the methods are supposed to throw. The rest is really nice!!.

Should we include some documentation of common methods that might be overwritten for performance? I'm thinking about the rename, isReadable, isCreatable, etc.

**Comment by PVince81** [comment]
Ahem ahem, there are no expected exceptions in this API... 😢 
Do you mean adding `StorageNotAvailableException` in every method signature ?

As for the docs, we could extend what I already posted to mention best practices when implementing an ext storage: https://github.com/owncloud/documentation/issues/2728

**Comment by jvillafanez** [comment]
> Do you mean adding `StorageNotAvailableException` in every method signature ?

At least that is being taken into account I think.

The problem is that if nobody tells you that you should throw an `StorageNotAvailableException` when you can't connect to the server, what I'd return is a false or any other value that could be interpreted as an error, which might cause some issues.

If there is no expected exceptions, throwing that exception is conceptually wrong 😢

**Comment by PVince81** [comment]
In the long term I hope we can redefine the storage APIs completely to be more simply and base them on our own concept (https://github.com/owncloud/core/issues/22388#issuecomment-262925510), not mimic PHP's system functions, which means exceptions for any errors.

@jvillafanez do you want me to add `StorageNotAvailableException` in the PHPDoc of the adapters for now ?

**Comment by jvillafanez** [comment]
If it's being handled I think it's good to have it documented

**Comment by PVince81** [comment]
@jvillafanez added here https://github.com/owncloud/core/pull/26675/commits/4208efcbfa4dfa0e823553ed97e4284f5ea89a6f

I took the opportunity to also add these on the `IStorage` interface where they were missing...

File: lib/public/Files/Storage/IStorage.php [hunk 1]
```diff
@@ -177,16 +190,19 @@ public function isDeletable($path);
...
 	 * @return bool
+	 * @throws StorageNotAvailableException if the storage is temporarily not available
 	 * @since 9.0.0
 	 */
 	public function isSharable($path);
 
 	/**
 	 * get the full permissions of a path.
 	 * Should return a combination of the PERMISSION_ constants defined in lib/public/constants.php
+	 * @throws StorageNotAvailableException if the storage is temporarily not available
```
Comment (jvillafanez) [inline lib/public/Files/Storage/IStorage.php]: duplicated below

**Comment by jvillafanez** [comment]
Other than that, 👍

File: lib/public/Files/Storage/IStorage.php (hunk 1 above)
Comment (PVince81) [inline lib/public/Files/Storage/IStorage.php]: my copy-paste skills have failed me 😢

**Comment by PVince81** [comment]
Fixed and squashed

**Pull request merged**